### ISSUE TITLE
Merge Node.js/io.js and browser JavaScript clients

### ISSUE THREAD

**Issue by vvo** [algolia/algoliasearch-client-js#46]
## tl;dr;

We currently have two JavaScript-like clients:
- Node.js: [npm/algolia-search](https://www.npmjs.com/package/algolia-search), [algolia/algoliasearch-client-node](https://github.com/algolia/algoliasearch-client-node)
- Browser: [npm/algoliasearch](https://www.npmjs.com/package/algoliasearch), [algolia/algoliasearch-client-js](https://github.com/algolia/algoliasearch-client-js)

Merge them in only one client: `algoliasearch`.

## Boring stuff

We can have [isomorphic](http://isomorphic.net/) JavaScript! Let's use the same JavaScript client on both Node.js and the browser.

As the request strategies are a little bit different between Node.js (keepalive using [node-modules/agentkeepalive](https://github.com/node-modules/agentkeepalive)) and the browser (fallback to JSONP, non relevant for Node.js), there will be some code differences.

Luckily, using browserify we are able to swap out requires when in the browser.

### First steps
- [x] Create V~~2~~1 branch in the Node.JS client, patches will be done in this branch
- [x] first prototype exposing AlgoliaSearch JavaScript client in Node.JS
- [x] make testing stack handles the Node.JS client

### Behavior
- [x] ~~All indexing requests must use a timeout of 15 times the requestTimeout given (so 2s search timeout, 30s indexing timeout)~~
- [x] Set default timeout to 3s. Timeout is an inactivity timeout as per [Node.JS request.setTimeout](https://nodejs.org/api/http.html#http_request_settimeout_timeout_callback)
- [x] All indexing requests must not use the dsnHost in both browser and node
- [x] Ensure headers must not be mutable between two requests calls (add a test)
- [x] Keepalive should be activated by default, it will not be possible to give another agent for now. It's not a requested feature, it should just works™
- [x] send x-user-agent as part of the url containing "Algolia for Node.JS/io.jsNode.js $NODEJSVERSION $MODULEVERSION" in both browser and node
- [x] let the user use only http if he wants to
- [x] remove any reference to document/window in AlgoliaSearch.js
- [x] provide a .destroy method in nodejs that will destroy the http Agent (and exiting the process)
- [x] implement custom headers: _addHeadersRateLimit, _addHeadersSecuredAPIKey

### Methods
- [x] **Implement** some Node.JS client methods not implemented in the JavaScript client
 - [x] index.getObjects
 - [x] index.deleteObjects
 - [x] index.deleteByQuery

- [x] **Implement and rework** some methods
  - [x] **rename** client.setTimeout in client.setRequestTimeout
  - [x] client.multipleQueries
    - [x] **Deprecate** client.startQueriesBatch, client.addQueryInBatch, client.sendQueriesBatch (already implemented in JS V3)
  - [x] client|index.addUserKey, client|index.addUserKeyWithValidity, client|index.addUserKeyWithValidityAndIndexes, client|index.updateUserKey, client|index.updateUserKeyWithValidity, client|index.updateUserKeyWithValidityAndIndexes => can be solve by only client|index.addUserKey (POST) and client|index.updateUserKey (PUT)
    - [x] **deprecate** client|index.addUserKeyWithValidity (already implemented in JS V3)
    - [x] **remove** client|index.addUserKeyWithValidityAndIndexes, client|index.updateUserKey, client|index.updateUserKeyWithValidity, client|index.updateUserKeyWithValidityAndIndexes (not implemented in JS V3)

- [x] **Implement only for node** some functions that cannot be ported to browsers:
  - [x] generateSecuredApiKey, it uses crypto functions, too heavy for browser
  - [x] allow passing functions to be attached to the prototype for non portable browser functions like 
  - [x] create default empty functions throwing when not implemented in the browser with a clear message
  - [x] remove ClassToDerive feature, must be done by the user
  - [x] disable cache in the Node.JS client on search and multiple queries, add params to createAlgoliaSearch
   - [x] client.useSecuredAPIKey, client.disableSecuredAPIKey
   - [x] client.enableRateLimitForward, client.disableRateLimitForward

### Parse build (src/server/builds/parse)
- [x] ~~requirable via `require('algoliasearch/Parse')`, but build in src/server/builds/parse.js~~
- [x] promise.delay will use nextTick()
- [x] ~~Inherit from Node.JS client, but use AlgoliaSearch constructor~~  
- [x] installable via wget/curl the main dist/file
- [x] build process using webpack (one big file to satisfy parse)
- [x] inherit from AlgoliaSearchServer, to share methods with the nodejs client

### Testing
- [x] test on travis-ci for Node .10 .12 io.js
- [x] add specific node.js request strategy: error, timeouts
- [x] test our node.js client is using keep-alive

### Examples
- [x] add node.js examples
- [x] add parse example

### Misc
- [x] [deprecate](https://docs.npmjs.com/cli/deprecate) npm/algolia-search (multiple latest versions)
- [x] update https://github.com/algolia/algoliasearch-client-node readme to display a warning + goto algoliasearch
- [x] update website
  - [x] clearly say that Node.JS and browser clients are the same and usable in both environments
  - [x] clearly say that some methods are not usable in the browser ()
- [x] create Node.JS migration guide
- [x] warn clients wanting to use the isomorphic JS client
- [ ] ~~ttAdapter belongs only to the browser build~~
- [x] get interesting info from Node.JS readme and set it in JavaScript readme
- [ ] ~~use and provide a modern parse setup with parse-cli parse-require~~ (impossible)
- [x] update website on parse

**Comment by vvo** [comment]
![meme](http://i.imgur.com/Lr2W5zV.jpg)

**Comment by redox** [comment]
:open_mouth: :100:

**Comment by vvo** [comment]
DONE :checkered_flag:
